yatta/src/desktop.rs

```diff
@@ -37,26 +37,137 @@ pub struct Desktop {
     pub foreground_window: Window,
-    pub gaps:              i32,
-    pub resize_step:       i32,
+    pub needs_recalc:      bool
+}
+
+impl Workspace {
+    pub fn verify(&mut self, windows: &mut Vec<Window>) -> bool {
+        let mut changed = false;
+        self.windows.retain(|window| {
+            let result = !windows.contains(window) && window.is_window();
+            if !result { changed = true; }
+            windows.push(*window);
+            result
+        });
+        changed
+    }
+}
+
+impl Default for Workspace {
+    fn default() -> Self {
+        Workspace {
+            foreground_window: Window::default(),
+            layout:            Layout::BSPV,
+            layout_dimensions: vec![],
+            windows:           vec![],
+            needs_recalc:      false
+        }
+    }
 }
 
 impl Display {
-    pub fn get_foreground_window(&mut self) {
-        self.foreground_window = Window::foreground();
+    pub fn get_workspace(&self) -> &Workspace {
+        self.workspaces.get(self.workspace_index).unwrap()
+    }
+
+    pub fn get_workspace_mut(&mut self) -> &mut Workspace {
+        self.workspaces.get_mut(self.workspace_index).unwrap()
+    }
+
+    pub fn create_workspace(&mut self, index: usize) {
+        while self.workspaces.len() <= index {
+            self.workspaces.push(Workspace::default());
+        }
+    }
+
+    pub fn get_current_windows(&self) -> &Vec<Window> {
+        &self.get_workspace().windows
+    }
+
+    pub fn get_current_windows_mut(&mut self) -> &mut Vec<Window> {
+        &mut self.get_workspace_mut().windows
+    }
+
+    pub fn get_layout(&self) -> &Layout {
+        &self.get_workspace().layout
+    }
+
+    pub fn get_layout_mut(&mut self) -> &mut Layout {
+        &mut self.get_workspace_mut().layout
+    }
+
+    pub fn get_layout_dimensions(&self) -> &Vec<Rect> {
+        &self.get_workspace().layout_dimensions
+    }
+
+    pub fn get_layout_dimensions_mut(&mut self) -> &mut Vec<Rect> {
+        &mut self.get_workspace_mut().layout_dimensions
+    }
+
+    pub fn needs_recalc(&self) -> bool {
+        self.get_workspace().needs_recalc
+    }
+
+    pub fn set_workspace(&mut self, index: usize) {
+        if index != self.workspace_index {
+            self.verify(&mut Vec::new());
+            self.create_workspace(index);
+            for window in self.get_current_windows_mut().iter_mut() {
+                window.hide();
+            }
+            self.workspace_index = index;
+            for window in self.get_current_windows_mut().iter_mut() {
+                window.restore();
+            }
+
+            let len_prev = self.get_current_windows().len();
+            self.get_current_windows_mut().retain(|x| x.is_window());
+            if self.get_current_windows().len() < len_prev || self.needs_recalc() {
+                self.calculate_layout();
+                self.apply_layout(None);
+            }
+            self.verify(&mut Vec::new());
```

Comment:
Currently when switching to a new workspace that has windows assigned to it, none of the associated windows will take the foreground focus and I need to use the mouse to focus one of them before I can start navigating again using the keyboard shortcuts.

Adding this snippet after the final `self.verify` call fixes this for me, always defaulting to giving focus to the first window on the workspace being switched to (this seems like a sensible default to me, but again I've never really been a heavy use of workspaces)

```
if let Some(window) = self.get_current_windows().first() {
    window.set_foreground()
}
```

Reply:
As I mostly use the mouse for managing windows, I didn't really notice this, I will add this fix when I'm on my computer again, if you haven't by then. Just FYI you can add commits to my pull request too by just pushing to the master branch of my fork (by default, maintainers of the original repository get commit access to the fork when a pull request is made).

Reply:
Also, better behavior might be to give back focus to the previously focused window.

Reply:
As it turns out, it isn't that easy, as hwnd IDs change every time a window is hidden.

Reply:
Oh, now I tried to do it by comparing titles, but as it turns out the yattac cmd window that appears for a split second is the foreground window when switching away :(

Reply:
I have a little bit of ab implementation problem: Autohotkey's ", Hide" doesn't seem to completely hide yattac's window, so when switching away from the current workspace yattac is always the last focused window.

In the latest commit, I modified yattac to not have a window at all, but this also means that commands like yattac help have no output at all.

Reply:
Also, as the hwnd's ID changes when hidden, I have to store the last focused window's title as a string, which could lead to issues when two windows have the same title.

Reply:
I've just tried running the latest build with these changes and I'm getting some funny behaviour where, although the `#![windows_subsystem = "windows"]` directive is supposed to suppress windows, now whenever `yattac` commands are getting run when I refresh my AHK config, a blue Powershell window is popping up every time. 

Once the key bindings have been set properly, I'm also not able to change focus reliably anymore (not too sure how to explain this, but if there is window beneath my focused window, focusing to the next window down will register an event but not update the display state, and when trying to swap windows up/down/left/right, one window will move but the other will not). Reverting back to an earlier build seems to solve these issues.

In light of the extra complexity of trying to revert back to the previously focused window on the last workspace, I'm happy to go with the more naive approach of defaulting to the 0th window in the list for each workspace upon switch, and opening a separate issue to track state across workspaces which could cover this and any other state-related refinements.

Reply:
This simpler approach indeed seems good enough to me for this pull request, I removed the windows-subsystem directive and implemented your behavior.